dspy/functional/typed_predictor_signature.py

```diff
@@ -0,0 +1,130 @@
+from typing import Annotated, Type, Union
+from typing import get_origin, get_args
+
+from pydantic import BaseModel
+from pydantic_core import PydanticUndefined
+from pydantic.fields import FieldInfo
+
+import dspy
+from dspy import InputField, OutputField, Signature
+
+
+class TypedPredictorSignature:
```

Comment:
Starting to leave notes here. FIrst thought: we don't want to change the normal interface of DSPy, e.g. no distinct classes for all inputs and for all outputs.

Reply:
Agreed